Updating ROC visualization

gwaybio opened the pull request:
Previous figure used loess without standard errors to visualize each of the training/testing partitions across each fold for each random seed. We were losing a lot of information by not plotting the distribution of ROC curves for each iteration.

In this PR, I plot each individual classifier (classifier for each random seed and each cross validation fold) in the ROC curve with loess smoothing.

Here is an example output: [example.pdf](https://github.com/greenelab/nf1_inactivation/files/607598/GBM_elasticnet_roc_outputalpha_0.15_l1ratio_0.1_roc.pdf)

gwaybio committed:
ignore ipynb files

gwaybio committed:
update classifier script

update saving cohens d files logic and importantly, saving intermediate ROC output files for better visualization

gwaybio committed:
update roc visualization

gwaybio committed:
remove comment arg in viz_roc

dhimmel commented:
Beautiful plots but they'll drive the reader crazy.

First, ROC curves are never decreasing -- it doesn't make sense that as you increase the False Positive Rate, the True Positive Rate could go down. Is there a reason these don't look like ROC curves?

gwaybio commented:
> First, ROC curves are never decreasing -- it doesn't make sense that as you increase the False Positive Rate, the True Positive Rate could go down. Is there a reason these don't look like ROC curves?

It all has to do with the choice to use [loess smoothing](http://www.statisticshowto.com/lowess-smoothing/) to visualize the distribution of points in the graph. The main reason I decided to use this visualization is because of the extremely low number of true positives in each test set. Essentially, there are only two or three in each fold. I think this demonstrates what I'm saying better:

![roc_curve_strange_pattern](https://cloud.githubusercontent.com/assets/7340421/20565054/f3c2dfc2-b15d-11e6-9cac-54b98b9dbecc.png)

Shown is one iteration of the classifier for both a `geom_step()` and `geom_line(stat=smooth)` with the points given. I am not sure if there is a better way to show this...

Also, if I plot all ROC curves with `geom_step()` it looks pretty bad!

![full_example_steps](https://cloud.githubusercontent.com/assets/7340421/20565800/eaaf2528-b160-11e6-9231-243ae00b137b.png)

gwaybio commented:
also tagging @cgreene for additional thoughts

cgreene commented:
If you want an aggregate over the curves, can you plot the median at each false positive rate? That would address @dhimmel's point (ROC curves don't go down). I think it also allows you to show what you're going for - namely the performance of the aggregate.

You could also throw 95% bootstrap confidence intervals on it. Using bootstrap would help b/c you avoid cases where the TPR goes above 1 (which loess doesn't prevent either).

dhimmel commented:
> If you want an aggregate over the curves, can you plot the median at each false positive rate?

Agree, but think @gwaygenomics should try the mean rather than median. This will get him that curve he's looking for.

gwaybio committed:
aggregate ROC over mean TP

gwaybio commented:
Thanks for the comments @dhimmel and @cgreene! Here is the current figure:

![gbm_elasticnet_roc_outputalpha_0 15_l1ratio_0 1_test](https://cloud.githubusercontent.com/assets/7340421/20679328/71fade4a-b568-11e6-96ab-36b889f2b24e.png)

> If you want an aggregate over the curves, can you plot the median at each false positive rate? That would address @dhimmel's point (ROC curves don't go down).

> Agree, but think @gwaygenomics should try the mean rather than median. This will get him that curve he's looking for.

Both the mean and median figures appear almost identical - I went with the mean value here. 

> I think it also allows you to show what you're going for - namely the performance of the aggregate.
You could also throw 95% bootstrap confidence intervals on it. Using bootstrap would help b/c you avoid cases where the TPR goes above 1 (which loess doesn't prevent either).

In the end, I decided to go with plotting each iteration separately fitting each with a simple `geom_line()`. I think the figure shows the heterogeneity of the splits, the caveats of small sample size, and the need to implement the model as an ensemble.

I will also need to edit the figure legend and respond the reviewer accordingly.

dhimmel commented:
This wasn't what I was thinking. I was thinking taking the average across all of the ROC curves. So you'd have a single test and a single train curve. I see the benefit of seeing the distribution of ROC curves, but I think it's outweighed by the fact that it will confuse most readers.

See http://stats.stackexchange.com/a/187003/74908

gwaybio commented:
> So you'd have a single test and a single train curve.

Got it - Yeah, I was wrestling with how to display this best. I definitely want so show heterogeneity but do not want to confuse the reader. If I add a reference to the aggregate model (or, how the ensemble model is expected to behave)(darker lines) in the legend and in the text I think this is a good compromise? What do you think?

![test_roc_agg](https://cloud.githubusercontent.com/assets/7340421/20683592/9d8468c6-b57a-11e6-8325-9ec040a196a1.png)

dhimmel commented:
If you take the average of all ROC curves in https://github.com/greenelab/nf1_inactivation/pull/10#issuecomment-262530121 can't you get a more continuous curve (with more than 4 points for testing)?.

> If I add a reference to the aggregate model (or, how the ensemble model is expected to behave)(darker lines) in the legend and in the text I think this is a good compromise? What do you think?

Yeah aggregate is a big improvement from before

gwaybio committed:
plot main aggregate curve

gwaybio commented:
> If you take the average of all ROC curves in #10 (comment) can't you get a more continuous curve (with more than 4 points for testing)?

Unfortunately, no. There are only ever 2 or 3 true positives in the test set and therefore, at any threshold you either have a 0, 0.3, 0.6, or 1 tpr. 

The most recent figure I added here is built in with _two_ aggregate means.

1. The mean of false positive rates within a single classifier. See a single iteration here:
![single iteration mean](https://cloud.githubusercontent.com/assets/7340421/20565054/f3c2dfc2-b15d-11e6-9cac-54b98b9dbecc.png)
2. The aggregate mean across all iterations (most recent fig)

gwaybio committed:
reduce alpha of iteration lines

gwaybio committed:
update .gitignore for merge issue

gwaybio commented:
Ok, thanks @dhimmel - here are the two plots we just discussed:

![geom_step_vs_geom_line](https://cloud.githubusercontent.com/assets/7340421/20723867/2d435fca-b63a-11e6-98ac-e539928878da.png)

The plot on the left is made with `geom_step()` and the plot on the right is `geom_line()`.

We also discussed how to potentially smooth the aggregate curve from the step wise curve. Essentially, the plot on the right _is_ the smoothed line since there are only ever horizontal and vertical step-wise lines in the ROC curve. I think I will go with the one on the left - what do you think? are there any other tweaks that would better get the message across?

gwaybio commented:
thanks for input and suggestions @dhimmel and @cgreene - this plot turned out to be a beast to summarize succinctly! Based on your comments, I think I will go with the [above plot on the right](https://github.com/greenelab/nf1_inactivation/pull/10#issuecomment-263662730).

@dhimmel your intuition was correct, here is what the plot looks like if I average across both TP and FP rates. It essentially mirrors the line plot above but is more complicated and would definitely frustrate a reader.

![average_tpr](https://cloud.githubusercontent.com/assets/7340421/20756493/e0e30138-b6e0-11e6-92d0-969982d157b2.png)

cgreene commented:
@gwaygenomics : Given all of this, I agree with the one that you've selected for the paper. The most recent plot might be what you'd want to use for a talk to an audience that will already get ROC curves (without the added line maybe?) if you want to talk about the importance of the ensemble. I'd suggest making both (also with a black background form) and choosing which one to use based on the audience.

gwaybio commented on scripts/nf1_classifier.py:
```diff
@@ -361,8 +366,10 @@ def plot_decision_function(score_df, partition, output_file):
 
-                            fpr, tpr, _ = roc_curve(y_test, y_score)
-                            fpr_t, tpr_t, _ = roc_curve(y_train, y_score_train)
+                            fpr, tpr, _ = roc_curve(y_test, y_score,
```
@dhimmel - here is where i output fpr and tpr for each classification

dhimmel replied:
See [`numpy.interp`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.interp.html) and [this example](http://scikit-learn.org/stable/auto_examples/model_selection/plot_roc.html). You will want to choose say 1000 FPRs from 0 to 1 and then get the corresponding TRP for each ROC curve. Then it will be a simple averaging of TPRs for each FPR and you will have your beautiful plot.

gwaybio replied:
Not as simple as this makes it seem! After many hours of interpolating over the ROC lines and trying many different potential solutions I have determined that this problem does not work as well for these particular points as it does in the example!

I am going to go with the [line graph here](https://github.com/greenelab/nf1_inactivation/pull/10#issuecomment-263662730) because it does tell the whole story of the figure and fits with the audience of the journal.

At some point in the future if you come across a similar scenario, or want to work through this one with me, that would be great.

gwaybio committed:
removing superfluous functionality and args

gwaybio commented:
@dhimmel @cgreene good to merge? I'm looking to get this wrapped up soon

dhimmel commented:
> @dhimmel @cgreene good to merge? I'm looking to get this wrapped up soon

Yep, I'm willing to spend an hour tomorrow with you to try to fix the ROC curve. But if you're fine with it now, then so am I.

cgreene commented:
:+1:

On Sun, Dec 4, 2016, 9:06 PM Daniel Himmelstein <notifications@github.com>
wrote:

> @dhimmel <https://github.com/dhimmel> @cgreene
> <https://github.com/cgreene> good to merge? I'm looking to get this
> wrapped up soon
>
> Yep, I'm willing to spend an hour tomorrow with you to try to fix the ROC
> curve. But if you're fine with it now, then so am I.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenelab/nf1_inactivation/pull/10#issuecomment-264752247>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAhHszpQPFH8uVqhDtjE1l5JozrbnA_Kks5rE3G2gaJpZM4K58mk>
> .
>

gwaybio commented:
> Yep, I'm willing to spend an hour tomorrow with you to try to fix the ROC curve. But if you're fine with it now, then so am I.

Sounds good - it would be great to go over it with you. I don't think we should spend any more than an hour though because the current plot is sufficient and this has already eaten a significant chunk of time!

dhimmel commented:
> I don't think we should spend any more than an hour though because the current plot is sufficient and this has already eaten a significant chunk of time!

I think we did it in just under an hour. See https://gist.github.com/dhimmel/07d689d963fca817f112773ffcb2d0ff. Make sure to add 0, 0 and 1, 1.

gwaybio commented:
:tada: :tada: it worked! :tada: :tada: and it looks great! thanks so much

![dhimmel_the_great](https://cloud.githubusercontent.com/assets/7340421/20904199/900d02e2-bb0c-11e6-951f-af430315f932.png)

This was frustrating me for days! Thanks @dhimmel

gwaybio committed:
implement interpolation solution

gwaybio committed:
plot roc ensemble

dhimmel commented:
@gwaygenomics what changed between the individual ROC curves in https://github.com/greenelab/nf1_inactivation/pull/10#issuecomment-262530121 and https://github.com/greenelab/nf1_inactivation/pull/10#issuecomment-264992068. The former concurs with our interpolated average line, so I think there's a bug somewhere that needs to be tracked down.

I made a few small [modifications to the notebook](https://gist.github.com/dhimmel/07d689d963fca817f112773ffcb2d0ff/revisions) (specified first and last in numpy.interp), but I don't think those should make a difference.

cgreene commented:
Looks to me like a change b/w upper and lower areas to approximate sums to me. Potentially in commit 63a829ba3d6848c7da426d1da8f1ace33ccb0047 since it looks like it changed in the figure immediately following that.

gwaybio commented:
> Make sure to add 0, 0 and 1, 1.

Yep, added!

> what changed between the individual ROC curves in #10 (comment) and #10 (comment). The former concurs with our interpolated average line, so I think there's a bug somewhere that needs to be tracked down.

Nice catch - simple fix.

> Looks to me like a change b/w upper and lower areas to approximate sums to me.

Exactly - except the bug was [here](https://github.com/greenelab/nf1_inactivation/pull/10/commits/30702b35e642ed79f406d11ea00f75c895bda4d1#diff-85bd3ad1fbee223be5fa05e009bfdf83R104). Cool to track down exact moment of bug!

> I made a few small modifications to the notebook (specified first and last in numpy.interp), but I don't think those should make a difference.

Did not make a difference, but thanks for updating! I also updated code in most recent commit.

Here is the figure - thanks again for your persistence!

![test_fig](https://cloud.githubusercontent.com/assets/7340421/20927771/0a070af0-bb91-11e6-9b94-713b1ff42213.png)

ready for your review again @dhimmel - I'd be happy to discuss more if you'd like

gwaybio committed:
fix aggregate group bug

not plotting mean_fpr

gwaybio committed:
specifying left and right in interpolate
